### PR TITLE
Add example for distributed launcher

### PR DESCRIPTION
**Pull request by drdarshan**
Resolves PyTorch issue #35160 (https://github.com/pytorch/pytorch/issues/35160). The example shows how to structure a DDP application so it an be started via the distributed launcher script in several configurations: one process per GPU, one process per node or anything in between.

### CODE REVIEW

File: distributed/ddp/ProcessMapping.svg [hunk 1]
```diff
@@ -0,0 +1 @@
...1.875 1.828125l0 3.96875l-1.515625 0zm22.007812 -1.359375l0 1.359375l-7.578125 0q-0.015625 -0.515625 0.171875 -0.984375q0.28125 -0.765625 0.921875 -1.515625q0.640625 -0.75 1.84375 -1.734375q1.859375 -1.53125 2.515625 -2.421875q0.65625 -0.90625 0.65625 -1.703125q0 -0.828125 -0.59375 -1.40625q-0.59375 -0.578125 -1.5625 -0.578125q-1.015625 0 -1.625 0.609375q-0.609375 0.609375 -0.609375 1.6875l-1.453125 -0.140625q0.15625 -1.625 1.125 -2.46875q0.96875 -0.84375 2.59375 -0.84375q1.65625 0 2.609375 0.921875q0.96875 0.90625 0.96875 2.25q0 0.6875 -0.28125 1.359375q-0.28125 0.65625 -0.9375 1.390625q-0.65625 0.734375 -2.171875 2.015625q-1.265625 1.0625 -1.625 1.453125q-0.359375 0.375 -0.59375 0.75l5.625 0z" fill-rule="nonzero"/><path fill="#000000" d="m387.98566 102.88231l0 -11.453125l1.515625 0l0 10.09375l5.640625 0l0 1.359375l-7.15625 0zm8.5078125 -5.578125q0 -2.859375 1.53125 -4.46875q1.53125 -1.609375 3.953125 -1.609375q1.578125 0 2.84375 0.765625q1.28125 0.75 1.953125 2.109375q0.671875 1.34375 0.671875 3.0625q0 1.75 -0.703125 3.125q-0.703125 1.375 -2.0 2.09375q-1.28125 0.703125 -2.78125 0.703125q-1.609375 0 -2.890625 -0.78125q-1.265625 -0.796875 -1.921875 -2.140625q-0.65625 -1.359375 -0.6
```
Comment (mrshenli) [inline distributed/ddp/ProcessMapping.svg]: another option to upload a image is to post it in the comment on github, and use the github perm-link in the code.

File: distributed/ddp/README.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+```
+that in turn produces the following output
+```sh
+[262816] Initializing process group with: {'MASTER_ADDR': '127.0.0.1', 'MASTER_PORT': '29500', 'RANK': '0', 'WORLD_SIZE': '1'}
+[262816]: world_size = 1, rank = 0, backend=nccl
+[262816] rank = 0, world_size = 1, n = 8, device_ids = [0, 1, 2, 3, 4, 5, 6, 7]
+```
+
+# Conclusions and next steps
+As a writer of a distribued data parallel application, your code needs to be aware of two types of resources: the compute nodes and the GPUs within each node. The process of setting up bookkeeping to track how the set of GPUs is mapped to the processes of your application can be tedious and error-prone. We hope that by structuring your application as shown in this example and using the launcher, the mechanics of setting up distributed training will be significiantly simplified.
```
Comment (mrshenli) [inline distributed/ddp/README.md]: distribued -> distributed
significiantly -> significantly

File: distributed/ddp/example.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    print(f"[{os.getpid()}] Initializing process group with: {env_dict}")
+    dist.init_process_group(backend="nccl")
+    print(
+        f"[{os.getpid()}]: world_size = {dist.get_world_size()}, "
+        + f"rank = {dist.get_rank()}, backend={dist.get_backend()}"
+    )
+
+    # Explicitly setting seed to make sure that models created in two processes
+    # start from same random weights and biases.
+    torch.manual_seed(42)
```
Comment (mrshenli) [inline distributed/ddp/example.py]: We might be able to get rid of this as DDP broadcast model states in construction. Could you please verify below is sufficient to sync models?

https://github.com/pytorch/pytorch/blob/df8d6eeb19423848b20cd727bc4a728337b73829/torch/nn/parallel/distributed.py#L298-L303

File: distributed/ddp/ProcessMapping.svg (hunk 1 above)
Comment (drdarshan) [reply distributed/ddp/ProcessMapping.svg]: Thanks for the quick review! Any preference on how to handle the image? It's pretty small and sometimes I've seen the GitHub permalink be not very stable (the image just disappears).

File: distributed/ddp/README.md (hunk 2 above)
Comment (drdarshan) [reply distributed/ddp/README.md]: Thanks, fixed.

File: distributed/ddp/example.py (hunk 3 above)
Comment (drdarshan) [reply distributed/ddp/example.py]: You are right indeed! I've removed it from the example. BTW, this is a super-nice feature to broadcast the state from rank 0.

File: distributed/ddp/ProcessMapping.svg (hunk 1 above)
Comment (mrshenli) [reply distributed/ddp/ProcessMapping.svg]: Hey @jlin27, what's the preferred way to upload images to docs and examples?

Comment (drdarshan) [reply distributed/ddp/ProcessMapping.svg]: Hello @mrshenli and @jlin27 - I removed the image for now from the repo and replaced it with a GitHub permalink. Please let me know if this looks OK. Thank you again!
